src/bsp_types/mappings/build_target.rs

```diff
@@ -0,0 +1,111 @@
+use std::ops::Add;
+
+use cargo_metadata::camino::Utf8Path;
+
+use crate::bsp_types::basic_bsp_structures::*;
+
+impl BuildTargetCapabilities {
+    pub fn new() -> Self {
+        BuildTargetCapabilities {
+            can_compile: false,
+            can_test: false,
+            can_run: false,
+            can_debug: false,
+        }
+    }
+
+    pub fn enable_all(&mut self) {
+        self.can_compile = true;
+        self.can_test = true;
+        self.can_run = true;
+        self.can_debug = true;
+    }
+
+    pub fn enable_compile(&mut self) -> &mut Self {
+        self.can_compile = true;
+        self
+    }
+
+    pub fn enable_test(&mut self) -> &mut Self {
+        self.can_test = true;
+        self
+    }
+
+    pub fn enable_run(&mut self) -> &mut Self {
+        self.can_run = true;
+        self
+    }
+
+    pub fn enable_debug(&mut self) -> &mut Self {
+        self.can_debug = true;
+        self
+    }
+}
+
+fn tags_and_capabilities_from_cargo_kind(
+    cargo_target: &cargo_metadata::Target,
+) -> (Vec<BuildTargetTag>, BuildTargetCapabilities) {
+    let mut tags = vec![];
+    let mut capabilities = BuildTargetCapabilities::new();
+    cargo_target
+        .kind
+        .iter()
+        .for_each(|kind| match kind.as_str() {
+            "lib" => {
+                tags.push(BuildTargetTag::Library);
+                capabilities.enable_compile().enable_test().enable_debug();
+            }
+            "bin" => {
+                tags.push(BuildTargetTag::Application);
+                capabilities.enable_all();
+            }
+            "example" => {
+                tags.push(BuildTargetTag::Application);
+                capabilities.enable_compile().enable_run().enable_debug();
+            }
+            "test" => {
+                tags.push(BuildTargetTag::Test);
+                capabilities.enable_compile().enable_run().enable_debug();
+            }
+            "bench" => {
+                tags.push(BuildTargetTag::Benchmark);
+                capabilities.enable_compile().enable_run().enable_debug();
+            }
+            "custom-build" => {
+                todo!()
+            }
+            _ => (),
+        });
+
+    (tags, capabilities)
+}
+
+fn discover_dependencies(_path: &Utf8Path) -> Vec<BuildTargetIdentifier> {
+    vec![] //todo
+}
+
+impl From<&cargo_metadata::Target> for BuildTarget {
+    fn from(cargo_target: &cargo_metadata::Target) -> Self {
+        let (tags, capabilities) = tags_and_capabilities_from_cargo_kind(cargo_target);
+
+        let mut base_directory = cargo_target.src_path.clone();
+        base_directory.pop();
```

Comment:
czemu zakładacie że to zawsze zadziała?

Reply:
Wyszedłem tutaj z założenia, że cargo metadata poda nam zawsze poprawną ścieżkę do pliku, która dodatkowo będzie miała ścieżkę do rodzica. Stąd nie bardzo widzę sytuacji, w której jest to podatne na błędy.

Reply:
to napisz to proszę w komentarzu